### PR TITLE
Fix local multi-subscriber thread broadcast (EventEmitterPubSub starvation)

### PR DESCRIPTION
**Pull request by mbenhamd**
## Summary

Fixes the long-standing **local multi-subscriber broadcast regression** in `thread-stream-runtime.ts` — the root cause behind ~7 failing/hanging `agent-signals` tests (documented as a known pre-existing issue in PR #200) and the broader sibling of the approval-marker bug fixed there.

### Root cause
`#prepareBroadcastSource` early-returned for the default `EventEmitterPubSub`, so all local same-runtime thread subscribers read the run record's **single shared `output.fullStream`**. A `ReadableStream` has one reader: the first consumer drained it and every other subscriber starved (tests hung). A coupled regression: the idle `behavior:'persist'` branch never broadcast — `#persistAndBroadcastIdleSignal` was dead code, so persisted signals (e.g. notification summaries) never reached subscribers.

### Fix (mirrors upstream's multicast model)
- `#withLocalBroadcastStream`: drains the source **exactly once** into a shared buffer; each subscriber gets an independent fan-out stream replayed from it (`createSubscriberStream` on the run record, consumed by `subscribeToThread`)
- The drain starts **eagerly at registration** — approval-suspension marker detection lives in the drain loop, so caller-only runs (no subscribers) still mark correctly and are never finalized-as-completed (review-found blocker, fixed + regression-tested)
- One-shot own-property `fullStream` outputs are captured up front and redefined as a multicast replay accessor — the caller reads through the shared buffer instead of racing the drain
- Idle-persist branch rewired to `#persistAndBroadcastIdleSignal` (persists once, broadcasts once)

### Validation
- The previously hanging multi-subscriber tests now pass and terminate (~tens of ms): multi-subscriber delivery, multicast-alive, notification summaries ×2, persisted idle signal, future-thread-run fan-out, no-replay-to-late-subscribers
- New regression test: caller-only approval-suspended run retains its record and resumes cleanly
- Gates: approval suites, tool-call-step, stream-until-idle, harness request-context + events (71), agent-channels (44), core `tsc` 0
- Two Codex review passes; all findings addressed (replay semantics, backpressure isolation, buffer lifecycle vs upstream model, ordering, own-property semantics)

### Known pre-existing failures NOT addressed (verified identical on the pre-fix tip)
`fans out sequential idle signal runs` (signal data-shape in idle-wake injection), `queues a signal from another agent`, `converts signals between DB/LLM/data part formats`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## ELI5

This PR fixes a "starvation" bug where multiple subscribers trying to read messages from the same local thread broadcast would hang. The problem was they were all reading from a single shared stream that could only be drained once—the first reader would consume everything, and the others would wait forever. The fix creates a shared buffer that drains the stream once, then gives each subscriber their own independent tap to read from that buffer, so everyone can read without blocking each other.

---

## Problem

A regression in the local (same-process) multi-subscriber broadcast mechanism caused approximately 7 `agent-signals` tests to hang. The root cause: when multiple subscribers tried to read from the same thread run's output stream via `EventEmitterPubSub`, they all received a reference to a single shared `ReadableStream`. Since a `ReadableStream` can only be drained once, the first consumer would exhaust it, causing all other subscribers to starve indefinitely.

Additionally, the idle signal persistence logic was broken—the "persist" behavior never actually broadcasted idle signals to subscribers.

### Main Fix: Multicast Subscriber Stream Factory

Modified `thread-stream-runtime.ts` to introduce a multicast pattern that:

1. **Drains once into a shared buffer**: `#withLocalBroadcastStream` consumes the source output exactly once into a shared buffer, preventing multiple readers from competing for the same stream.

2. **Creates independent fan-out streams per subscriber**: Each subscriber gets an independent replay stream (`createSubscriberStream`) that reads from the shared buffer, so they can consume at their own pace without blocking each other.

3. **Eager drain at registration**: The drain starts immediately when a run is registered, not when the first subscriber connects. This ensures approval-suspension markers are detected and recorded even if there are currently zero subscribers (preventing caller-only runs from being incorrectly finalized without retaining output).

4. **Captures and replays fullStream outputs**: One-shot `fullStream` outputs are captured up front and redefined as multicast replay accessors, allowing callers to read through the shared buffer without racing the drain.

### Secondary Fix: Idle Signal Broadcasting

Rewired the idle "persist" signal routing to call `#persistAndBroadcastIdleSignal`, ensuring that persisted idle signals are properly broadcast to all subscribers (not just persisted silently).

## Changes

**`packages/core/src/agent/thread-stream-runtime.ts`** (+172/-40)
- `#prepareBroadcastSource` now returns a multicast `createSubscriberStream` factory for local runs (replacing inline approval detection logic)
- `AgentThreadRunRecord` optionally carries `createSubscriberStream` for per-run subscriber stream creation
- Thread subscriptions now read from `run.createSubscriberStream?.()` with fallback to `run.output.fullStream`
- Idle "persist" behavior now calls `#persistAndBroadcastIdleSignal` to broadcast persisted signals

**`packages/core/src/agent/__tests__/agent-signals.test.ts`** (+95/-0)
- Added regression test: verifies that approval-suspended runs without active subscribers still retain their output and record suspension state, allowing them to be resumed later
- Test ensures the fix handles the edge case of caller-only runs (no `subscribeToThread` consumers) correctly

## Validation

- Previously hanging multi-subscriber tests now pass and terminate correctly
- New regression test ensures the caller-only approval-suspension scenario works as expected
- Multiple test suites and type checks reported passing
- Known pre-existing failures remain unchanged

### CODE REVIEW

File: packages/core/src/agent/thread-stream-runtime.ts [hunk 1]
```diff
@@ -1893,7 +2013,19 @@ export class AgentThreadStreamRuntime {
...
-        const persisted = this.#persistSignal(
+        // Persist the signal AND broadcast it to thread subscribers (upstream
+        // parity). The fork previously only persisted, so idle-persist signals —
+        // e.g. low-priority notification summaries dispatched via
+        // `ifIdle: { behavior: 'persist' }` — never reached same-runtime
+        // subscribers. #persistAndBroadcastIdleSignal persists first, then emits
+        // a synthetic start/data/finish run through the (multicast) broadcast
+        // machinery without waking the agent.
+        key ??= this.#threadKey(resourceId, threadId);
+        const persisted = this.#persistAndBroadcastIdleSignal(
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/agent/thread-stream-runtime.ts:2024]: Fixed in follow-up PR #204 — run-completed now gates on registration + stream-part broadcast settling (async-pubsub consumers can no longer see completion before the parts), and replay-stream errors are presence-tracked so falsy throws fail the stream instead of closing it cleanly.

File: packages/core/src/agent/thread-stream-runtime.ts [hunk 2]
```diff
@@ -260,44 +267,148 @@ export class AgentThreadStreamRuntime {
...
+          const source = ownFullStream ? capturedSource : (output.fullStream as any);
+          if (!source) return;
+          for await (const part of source) {
+            runtime.#markApprovalSuspendedFromPart(pubsub, runId, part);
+            parts.push(part);
+            wake();
+          }
+        } catch (caught) {
+          error = caught;
+        } finally {
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/agent/thread-stream-runtime.ts:327]: Fixed in follow-up PR #204 — run-completed now gates on registration + stream-part broadcast settling (async-pubsub consumers can no longer see completion before the parts), and replay-stream errors are presence-tracked so falsy throws fail the stream instead of closing it cleanly.
